spec/integration/application/filebucket_spec.rb

```diff
@@ -70,10 +70,114 @@
         filebucket.command_line.args << 'fac251367c9e083c6b1f0f3181'
         filebucket.run
       }.to output(a_string_matching(
-        %r{Debug: HTTP GET https:\/\/127.0.0.1:#{port}\/puppet\/v3\/file_bucket_file\/md5\/fac251367c9e083c6b1f0f3181 returned 200 OK}
+        %r{Debug: HTTP GET https:\/\/127.0.0.1:#{port}\/puppet\/v3\/file_bucket_file\/md5\/fac251367c9e083c6b1f0f3181\?environment\=production returned 200 OK}
       ).and matching(
         %r{something to store}
        )).to_stdout
     end
   end
+
+  context 'diff', unless: Puppet::Util::Platform.windows? || Puppet::Util::Platform.jruby? do
+    context 'using a remote bucket' do
+      it 'outputs a diff between a local and remote file' do
+        File.binwrite(backup_file, "bar\nbaz")
+
+        get_handler = -> (req, res) {
+          res['Content-Type'] = 'application/octet-stream'
+          res.body = 'foo'
+        }
+
+        server.start_server(mounts: {filebucket: get_handler}) do |port|
+          Puppet[:masterport] = port
+          expect {
+            filebucket.command_line.args += ['diff', 'fac251367c9e083c6b1f0f3181', backup_file, '--remote']
+            filebucket.run
+          }.to output(a_string_matching(
+            /[-<] ?foo/
+          ).and matching(
+            /[+>] ?bar/
+          ).and matching(
+            /[+>] ?baz/
+          )).to_stdout
+        end
+      end
+
+      it 'outputs a diff between two remote files' do
+        get_handler = -> (req, res) {
+          res['Content-Type'] = 'application/octet-stream'
+          res.body = "--- /opt/puppetlabs/server/data/puppetserver/bucket/d/3/b/0/7/3/8/4/d3b07384d113edec49eaa6238ad5ff00/contents\t2020-04-06 21:25:24.892367570 +0000\n+++ /opt/puppetlabs/server/data/puppetserver/bucket/9/9/b/9/9/9/2/0/99b999207e287afffc86c053e5693247/contents\t2020-04-06 21:26:13.603398063 +0000\n@@ -1 +1,2 @@\n-foo\n+bar\n+baz\n"
+        }
+
+        server.start_server(mounts: {filebucket: get_handler}) do |port|
+          Puppet[:masterport] = port
+          expect {
+            filebucket.command_line.args += ['diff', 'd3b07384d113edec49eaa6238ad5ff00', "99b999207e287afffc86c053e5693247", '--remote']
+            filebucket.run
+          }.to output(a_string_matching(
+            /[-<] ?foo/
+          ).and matching(
+            /[+>] ?bar/
+          ).and matching(
+            /[+>] ?baz/
+          )).to_stdout
+        end
+      end
+    end
+
+    context 'using a local bucket' do
+      let(:filea) {
+        f = tmpfile('filea')
+        File.binwrite(f, 'foo')
+        f
+      }
+      let(:fileb) {
+        f = tmpfile('fileb')
+        File.binwrite(f, "bar\nbaz")
+        f
+      }
+      let(:checksuma) { Digest::MD5.file(filea).to_s }
+      let(:checksumb) { Digest::MD5.file(fileb).to_s }
```

Comment:
We were talking about something with this, but I don't remember what it was...

Reply:
I think we were talking about how you could move the writing to a tmpfile to a `before` block to simplify the `let(:filea)` above, but what you have is good, it guarantees the file contents will be written as soon as `filea` is resolved.